wheatley/parsing.py

```diff
@@ -6,11 +6,53 @@
 import logging
 
 from wheatley.aliases import CallDef, JSON
+from wheatley.bell import Bell
 from wheatley.row_generation import RowGenerator
 from wheatley.row_generation.place_notation_generator import PlaceNotationGenerator
 from wheatley.row_generation.complib_composition_generator import ComplibCompositionGenerator, \
                                                                   PrivateCompError, InvalidCompError
 
+class StartRowParseError(ValueError):
+    """
+    An error thrown with a helpful error message when the user inputs a start row string that
+    cannot be parsed by the parser.
+    """
+
+    def __init__(self, start_row_string: str, message: str) -> None:
+        super().__init__()
+
+        self.start_row_string = start_row_string
+        self.message = message
+
+    def __str__(self) -> str:
+        return f"Error parsing start row '{self.start_row_string}': {self.message}"
+
+def parse_start_row(start_row: str) -> int:
+    """
+    Parses a start row written in the format "12345" to ensure all characters are valid bells
+    and there are no missing bells.
+    """
+    def exit_with_message(error_text: str) -> NoReturn:
+        """ Raise an exception with a useful error message. """
+        raise StartRowParseError(start_row, error_text)
+
+    max_bell: int = 0
+
+    if start_row is None:
+        return max_bell
+
+    for char in start_row:
+        try:
+            bell = Bell.from_str(char)
+        except ValueError as e:
+            exit_with_message(str(e))
+
+        max_bell = max(max_bell, bell.number)
+
+    if max_bell > len(start_row):
+        exit_with_message(f"Start row does not contain all bells from 1 to {max_bell}")
```

Comment:
I think this error message would be nicer if it were more specific and told us which bell(s) are missing - even just outputting the smallest missing bell would be better.

I have some more test cases for row parsing in [a different project](https://github.com/kneasle/3rd-yr-proj/blob/master/core/src/row.rs#L793-L858) which give an example of nice error messages (they are written in Rust though, so I'm not sure how readable they are).

Reply:
I have extended this to give the list of missing bells.
Seeing as at that point it's basically free I also added a check for dupe bells at this point (which can name the exact bell).

This now means the dupe bell error in helper should be unreachable, so hopefully, that helps with the comment below on that error.